File: src/bmsx/machine/devices/irq/irq_controller.ts
import { IO_IRQ_ACK, IO_IRQ_FLAGS } from '../../bus/io';
import { Memory } from '../../memory/memory';

export class IrqController {
	private _signalSequence = 0;

	public constructor(private readonly memory: Memory) {
		this.memory.mapIoWrite(IO_IRQ_ACK, this.onAckWrite.bind(this));
	}

	public get signalSequence(): number {
		return this._signalSequence;
	}

	public reset(): void {
		this._signalSequence = 0;
		this.memory.writeIoValue(IO_IRQ_FLAGS, 0);
		this.memory.writeIoValue(IO_IRQ_ACK, 0);
	}

	public postLoad(): void {
		this._signalSequence = 0;
		this.memory.writeIoValue(IO_IRQ_ACK, 0);
	}

	public pendingFlags(): number {
		return this.memory.readIoU32(IO_IRQ_FLAGS);
	}

	public raise(mask: number): void {
		const current = this.memory.readIoU32(IO_IRQ_FLAGS);
		const next = (current | (mask >>> 0)) >>> 0;
		this.memory.writeIoValue(IO_IRQ_FLAGS, next);
		if (next !== current) {
			this._signalSequence += 1;
		}
	}

	public acknowledge(mask: number): void {
		const ack = mask >>> 0;
		if (ack !== 0) {
			this.memory.writeIoValue(IO_IRQ_FLAGS, (this.memory.readIoU32(IO_IRQ_FLAGS) & ~ack) >>> 0);
		}
		this.memory.writeIoValue(IO_IRQ_ACK, 0);
	}

	private onAckWrite(): void {
		this.acknowledge(this.memory.readIoU32(IO_IRQ_ACK));
	}
}

File: src/bmsx_cpp/machine/devices/irq/irq_controller.cpp
#include "machine/devices/irq/irq_controller.h"
#include "machine/bus/io.h"

namespace bmsx {

IrqController::IrqController(Memory& memory)
	: m_memory(memory) {
	m_memory.mapIoWrite(IO_IRQ_ACK, this, &IrqController::onAckWriteThunk);
}

void IrqController::reset() {
	m_signalSequence = 0;
	m_memory.writeIoValue(IO_IRQ_FLAGS, valueNumber(0.0));
	m_memory.writeIoValue(IO_IRQ_ACK, valueNumber(0.0));
}

void IrqController::postLoad() {
	m_signalSequence = 0;
	m_memory.writeIoValue(IO_IRQ_ACK, valueNumber(0.0));
}

uint32_t IrqController::pendingFlags() const {
	return m_memory.readIoU32(IO_IRQ_FLAGS);
}

void IrqController::raise(uint32_t mask) {
	const uint32_t current = m_memory.readIoU32(IO_IRQ_FLAGS);
	const uint32_t next = current | mask;
	m_memory.writeIoValue(IO_IRQ_FLAGS, valueNumber(static_cast<double>(next)));
	if (next != current) {
		m_signalSequence += 1;
	}
}

void IrqController::acknowledge(uint32_t mask) {
	if (mask != 0u) {
		const uint32_t flags = m_memory.readIoU32(IO_IRQ_FLAGS) & ~mask;
		m_memory.writeIoValue(IO_IRQ_FLAGS, valueNumber(static_cast<double>(flags)));
	}
	m_memory.writeIoValue(IO_IRQ_ACK, valueNumber(0.0));
}

void IrqController::onAckWriteThunk(void* context, uint32_t, Value) {
	static_cast<IrqController*>(context)->onAckWrite();
}

void IrqController::onAckWrite() {
	acknowledge(m_memory.readIoU32(IO_IRQ_ACK));
}

} // namespace bmsx
